[Key Vault Admin] Docs and Samples (#11075)
Azure/azure-sdk-for-js: sdk/keyvault/keyvault-admin/README.md

Azure Key Vault is a service that allows you to encrypt authentication keys, storage account keys, data encryption keys, .pfx files, and passwords by using secured keys.
If you would like to know more about Azure Key Vault, you may want to review: [What is Azure Key Vault?][docs-overview]

The package `@azure/keyvault-admin` provides support for the administrative Key Vault tasks. It includes the backup and restore operations for whole Key Vault instances, and the role-based access control (RBAC) operations.

[Source code][package-gh] | [Package (npm)][package-npm] | [API Reference Documentation][docs] | [Product documentation][docs-service] | [Samples][samples]

## Getting started

### Install the package

Install the Azure Key Vault administration client library for JavaScript and TypeScript with [NPM][NPM]:

```PowerShell
npm install @azure/keyvault-admin
```

### Prerequisites

* An [Azure subscription][azure-sub].
* An existing Azure Key Vault. If you need to create an Azure Key Vault, you can use the [Azure CLI][azure-cli].
* If you are using this package in a Node.js application, then use Node.js 8.x or higher.

> See the final two steps in the next section for details on creating the Key Vault with the Azure CLI.

Reviewer: I would recommend linking to the right section from the second bullet point above which talks about creting key vault using azure cli instead of a separate note like this

Author: I believe I fixed this! Please review again.

* An existing Azure Key Vault. If you need to create an Azure Key Vault, you can use the [Azure CLI][azure-cli].
* Use [Node.js](https://nodejs.org/) 8.x or higher.

#### Getting Azure credentials

Use the [Azure CLI][azure-cli] snippet below to create/get client secret credentials.

 * Create a service principal and configure its access to Azure resources:
    ```PowerShell
    az ad sp create-for-rbac -n <your-application-name> --skip-assignment
    ```
    Output:
    ```json
    {
        "appId": "generated-app-ID",
        "displayName": "some-app-name",
        "name": "http://some-app-name",
        "password": "random-password",
        "tenant": "tenant-ID"
    }
    ```
* Take note of the service principal objectId
    ```PowerShell
    az ad sp show --id <appId> --query objectId
    ```
    Output:
    ```
    "<your-service-principal-object-id>"
    ```
* Use the returned credentials above to set  **AZURE_CLIENT_ID** (appId), **AZURE_CLIENT_SECRET** (password), and **AZURE_TENANT_ID** (tenant) environment variables:
    ```PowerShell
    AZURE_TENANT_ID=<tenant-ID>
    AZURE_CLIENT_ID=<generated-app-ID>
    AZURE_CLIENT_SECRET=<random-password>

>     ```

Reviewer: The above is not a powershell snippet. Either skip the code snippet entirely or follow the .Net readme where the powershell script is used

Author: I'll remove `PowerShell` from here!

Author: I'll remove this code snippet entirely.

# Azure Key Vault Administration client library for JavaScript

Azure Key Vault is a service that allows you to encrypt authentication keys, storage account keys, data encryption keys, .pfx files, and passwords by using secured keys.
If you would like to know more about Azure Key Vault, you may want to review: [What is Azure Key Vault?][docs-overview]

The package `@azure/keyvault-admin` provides support for the administrative Key Vault tasks. It includes the backup and restore operations for whole Key Vault instances, and the role-based access control (RBAC) operations.

[Source code][package-gh] | [Package (npm)][package-npm] | [API Reference Documentation][docs] | [Product documentation][docs-service] | [Samples][samples]

## Getting started

### Install the package

Install the Azure Key Vault administration client library for JavaScript and TypeScript with [NPM][NPM]:

```PowerShell
npm install @azure/keyvault-admin
```

### Prerequisites

To work with the Azure Key Vault Administration client, the following is necessary:

* An [Azure subscription][azure-sub].
* An existing Azure Key Vault. If you need to create an Azure Key Vault, you can use the [Azure CLI][azure-cli].

> * If you are using this package in a Node.js application, then use Node.js 8.x or higher.

Reviewer: I recall that none of the keys/secrets/certificates packages work in the browser. Does that hold true for the admin package as well? Because the above statement leads one to believe that this package works in both node and browser

Author: 🤔 ok

Author: Doing it!

#### Getting Azure credentials

Use the [Azure CLI][azure-cli] snippet below to create/get client secret credentials.

 * Create a service principal and configure its access to Azure resources:
    ```PowerShell
    az ad sp create-for-rbac -n <your-application-name> --skip-assignment
    ```
    Output:
    ```json
    {
        "appId": "generated-app-ID",
        "displayName": "some-app-name",
        "name": "http://some-app-name",
        "password": "random-password",
        "tenant": "tenant-ID"
    }
    ```
* Take note of the service principal objectId
    ```PowerShell
    az ad sp show --id <appId> --query objectId
    ```
    Output:
    ```
    "<your-service-principal-object-id>"
    ```
* Use the returned credentials above to set  **AZURE_CLIENT_ID** (appId), **AZURE_CLIENT_SECRET** (password), and **AZURE_TENANT_ID** (tenant) 

> environment variables. For example, you can use the [dotenv][dotenv] project and add a `.env` file to your project with hte following contents (for the full section of environment variables, see [Environment Variables](#environment-variables)):

Reviewer: Based on the discussion in https://github.com/Azure/azure-sdk-for-js/pull/11075#discussion_r485386025, I would refrain from talking about dotenv and sample.env files.

    ```PowerShell
    az keyvault show --hsm-name <your-key-vault-name>
    ```

#### Get or create a storage account

To generate Key Vault backups, you will need to point the [KeyVaultBackupClient][src-backup-client] to an existing Storage account.

To create a new Storage Account, you can use the [Azure Portal][storage-account-create-portal],
[Azure PowerShell][storage-account-create-ps], or the [Azure CLI][storage-account-create-cli].
Here's an example using the Azure CLI:

```Powershell
az storage account create --name MyStorageAccount --resource-group MyResourceGroup --location westus --sku Standard_LRS
```

#### Environment Variables

With a Key Vault and a Storage account, you can fill the rest of the environment variables in the `.env` file used by [dotenv][dotenv]. You can look for more information about other required environment variables in our [sample.env](./sample.env) file.

#### Create KeyVaultAccessControlClient

Once you've populated the **AZURE_CLIENT_ID**, **AZURE_CLIENT_SECRET** and **AZURE_TENANT_ID** environment variables and replaced **your-vault-url** 
with the above returned URI, you can create the [KeyVaultAccessControlClient][src-rbac-client]:

```ts

> const credentials = new DefaultAzureCredential();

Reviewer 2: do you also want to show the necessary imports in this snippet

const client = new KeyVaultAccessControlClient(vaultUrl, credentials);
```

#### Create KeyVaultBackupClient

Once you've populated the **AZURE_CLIENT_ID**, **AZURE_CLIENT_SECRET** and **AZURE_TENANT_ID** environment variables and replaced **your-vault-url** 
with the above returned URI, you can create the [KeyVaultBackupClient][src-backup-client]:

```ts

> const vaultUrl = `https://<MY KEY VAULT HERE>.vault.azure.net`;

Reviewer 2: nit: this snippet has this extra `vaultUrl` line compared to the previous snippet, can you also include this line in the previous snippet?

const credentials = new DefaultAzureCredential();
const client = new KeyVaultBackupClient(vaultUrl, credentials);
```

## Key concepts

### KeyVaultRoleDefinition

> A `KeyVaultRoleDefinition` is a collection of permissions. A role definition defines the operations that can be performed, such as read, write, 

Reviewer 2: question: does the name of an object matter a lot in JS? If not, it might be better to focus less on the name of the object, and more abstractly on what the concept is

and delete. It can also define the operations that are excluded from allowed operations.

KeyVaultRoleDefinitions can be listed and specified as part of a `KeyVaultRoleAssignment`.

### KeyVaultRoleAssignment. 

A `KeyVaultRoleAssignment` is the association of a KeyVaultRoleDefinition to a service principal. They can be created, listed, fetched individually, and deleted.

### KeyVaultAccessControlClient

A [KeyVaultAccessControlClient][src-rbac-client] provides both synchronous and asynchronous operations allowing for management of `KeyVaultRoleDefinition` and `KeyVaultRoleAssignment` objects.

### KeyVaultBackupClient

A [KeyVaultBackupClient][src-backup-client] provides both synchronous and asynchronous operations for performing full key backups, full key restores, and selective key restores.

### Long Running Operations

> The Azure SDK for TypeScript and JavaScript offers a common abstraction for long running operations through the package [core-lro][corelro]. In particular, the operations done by the [KeyVaultBackupClient][src-backup-client] may take as much time as needed by the Azure resources, requiring a client layer to keep track, serialize and resume the operations through the lifecycle of the programs that wait for them to finish.

Reviewer 3: it doesn't appear that `[corelro]` exists in the links below

Author: thank you!

npm install @types/node
```

You also need to enable `compilerOptions.allowSyntheticDefaultImports` in your tsconfig.json. Note that if you have enabled `compilerOptions.esModuleInterop`, `allowSyntheticDefaultImports` is enabled by default. See [TypeScript's compiler options handbook][compiler-options] for more information.

### Authenticate the client

In order to control permissions to the Key Vault service, or to generate and restore backups of a specific Key Vault, you'll need to create either an instance of the [KeyVaultAccessControlClient][src-rbac-client] class, or an instance of the [KeyVaultBackupClient][src-backup-client] class, respectively.

> In both cases, you'll need a **vault URL**, which you may see as "DNS Name" in the portal,  and **client secret credentials (client id, client secret, tenant id)** to instantiate a client object.
> 
> Client secret credential authentication is being used in this getting started section but you can find more ways to authenticate with [Azure identity][azure-identity]. To use the [DefaultAzureCredential][DAC] provider shown below, or other credential providers provided with the Azure SDK, you should install the [@azure/identity][identity-npm] package:

Reviewer: ```suggestion
In both cases, you'll need a **vault URL**, which you may see as "DNS Name" in the portal,  and a credential object from the [@azure/identity](identity-npm) package which is used to work with Azure Active Directory.

In the below example, we are using a **client secret credentials (client id, client secret, tenant id)**,  but you can find more ways to authenticate with [Azure Identity][azure-identity]. To use the [DefaultAzureCredential][DAC] provider shown below, or other credential providers provided with the Azure SDK, you should install the [@azure/identity][identity-npm] package:
```

(replace `<your-resource-group-name>` and `<your-key-vault-name>` with your own, unique names and `<your-service-principal-object-id>` with the value from above):
    ```
    az keyvault create --hsm-name <your-key-vault-name> --resource-group <your-resource-group-name> --administrators <your-service-principal-object-id> --location <your-azure-location>
    ```

* Use the above mentioned Azure Key Vault name to retrieve details of your Vault which also contains your Azure Key Vault URL:
    ```PowerShell
    az keyvault show --hsm-name <your-key-vault-name>
    ```

#### Get or create a storage account

To generate Key Vault backups, you will need to point the [KeyVaultBackupClient][src-backup-client] to an existing Storage account.

To create a new Storage Account, you can use the [Azure Portal][storage-account-create-portal],
[Azure PowerShell][storage-account-create-ps], or the [Azure CLI][storage-account-create-cli].
Here's an example using the Azure CLI:

```Powershell
az storage account create --name MyStorageAccount --resource-group MyResourceGroup --location westus --sku Standard_LRS
```

> #### Environment Variables

Reviewer: I am not convinced that having this section here on env variables, sample.env and dotenv package in the readme is useful for the getting started scenario. Those copy/pasting code from the readme file should be able to do the bare minimum to get things started.

I see the merit of this in the Examples section because we are linking to samples that expect many env variables to be set.

Author: I'll remove it, and I'll link to the samples readme better!

Author: Here's what I'm placing in the examples:

> ## Examples
> 
> We have provided readmes with the detailed steps to run our samples, both for [JavaScript](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/javascript/README.md) and [TypeScript](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/README.md). Direct links to the specific samples follow:
>
> ...

const credentials = new DefaultAzureCredential();

const vaultUrl = `https://<MY KEY VAULT HERE>.vault.azure.net`;
const client = new KeyVaultBackupClient(vaultUrl, credentials);
```

## Key concepts

### KeyVaultRoleDefinition

A Role Definition is a collection of permissions. A role definition defines the operations that can be performed, such as read, write, 
and delete. It can also define the operations that are excluded from allowed operations.

Role definitions can be listed and specified as part of a Role Assignment.

### KeyVaultRoleAssignment. 

A Role Assignment is the association of a Role Definition to a service principal. They can be created, listed, fetched individually, and deleted.

### KeyVaultAccessControlClient

A [KeyVaultAccessControlClient][src-rbac-client] provides both synchronous and asynchronous operations allowing for management of Role Definitions (instances of `KeyVaultRoleDefinition`) and Role Assignments (instances of `KeyVaultRoleAssignment`).

### KeyVaultBackupClient

A [KeyVaultBackupClient][src-backup-client] provides both synchronous and asynchronous operations for performing full key backups, full key restores, and selective key restores.

> ### Long Running Operations

Reviewer: LRO is an abbreviation we use internally and I don't expect end users to be familiar with it. 

```suggestion
### Long running operations
```

Author: The rest of the titles are upper-cased like this one 🤔

> The Azure SDK for TypeScript and JavaScript offers a common abstraction for long running operations through the package [core-lro][core-lro]. In particular, the operations done by the [KeyVaultBackupClient][src-backup-client] may take as much time as needed by the Azure resources, requiring a client layer to keep track, serialize and resume the operations through the lifecycle of the programs that wait for them to finish.

Reviewer: ```suggestion
The Azure SDK for TypeScript and JavaScript offers a common abstraction for long running operations through the package [@azure/core-lro][core-lro]. In particular, the operations done by the [KeyVaultBackupClient][src-backup-client] may take as much time as needed by the Azure resources, requiring a client layer to keep track, serialize and resume the operations through the lifecycle of the programs that wait for them to finish.
```

Author: I think we should mention it. The Poller methods aren't part of the client's public API. The public API file imports from core-lro:

https://github.com/Azure/azure-sdk-for-js/pull/11009/files#diff-d1d574cf276744fd53761f9a30b41d49R9-R10

Reviewer: Well, the question is whether the user needs to import the core-lro package in order to use these long running operations. If no, then I would recommend not mentioning it here.

Even if we do want to mention it, I would put the line on it in the end and not attract too much attention to it. We want to draw attention to the fact that there are long running operations in this package and these methods are different from the others. The core lro package is implementation details

```suggestion
The operations done by the [KeyVaultBackupClient][src-backup-client] may take as much time as needed by the Azure resources, requiring a client layer to keep track, serialize and resume the operations through the lifecycle of the programs that wait for them to finish. This is done via a common abstraction through the package [@azure/core-lro][core-lro]. 
```

Author: > Well, the question is whether the user needs to import the core-lro package in order to use these long running operations. If no, then I would recommend not mentioning it here.

No, they don't need to import this package.

I like that update! I'll add it.

The [KeyVaultBackupClient][src-backup-client] offers three methods that execute Long Running Operations:

- `beginBackup`, starts generating a backup of an Azure Key Vault on the specified Storage Blob account.
- `beginRestore`, starts restoring all key materials using the SAS token pointing to a previously stored Azure Blob storage backup folder.
- `beginSelectiveRestore`, starts restoring all key versions of a given key using user supplied SAS token pointing to a previously stored Azure Blob storage backup folder.

> All of the Long Running Operation methods return a a Long Running Operation poller that allows you to wait indefinitely until the Key Vault selective restore is complete. More information is available on the examples below.

Reviewer: LRO is an abbreviation we use internally and I don't expect end users to be familiar with it.
Also, here we begin by referring to lro methods in general, but then pivot to the restore method.

```suggestion
The methods that begin long running operations return a poller that allows you to wait indefinitely until the operation is complete. More information is available on the examples below.
```

Author: I like this!

## Examples

> The following section provides several code snippets using the `client` created above for either [access control](#create-keyvaultaccesscontrolclient) or [backup](#create-KeyVaultBackupClient) clients, covering some of the most common Azure Key Vault access control related tasks:

Reviewer: > covering some of the most common Azure Key Vault access control related tasks

Since both access control and backup clients are referred to in this section, it is odd that we say that the section covers the most common "access control related tasks"

> using the `client` created above

The reference to "client created above" is not needed and forces one who lands on this section directly to then scroll to another section. Since the samples being linked from here have the client creation code, we do not need to link to client creation ourselves.

```suggestion
The following section provides several code snippets using the `KeyVaultAccessControlClient` for access control tasks and `KeyVaultBackupClient` for backup/restore tasks.
```

Author: I like this! thank you 🌞

Author: Here's what I'm placing in the examples:

> ## Examples
> 
> We have provided readmes with the detailed steps to run our samples, both for [JavaScript](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/javascript/README.md) and [TypeScript](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/README.md). Direct links to the specific samples follow:
>
> ...

Feedback appreciated!

### Long running operations

The Azure SDK for TypeScript and JavaScript offers a common abstraction for long running operations through the package [core-lro][core-lro]. In particular, the operations done by the [KeyVaultBackupClient][src-backup-client] may take as much time as needed by the Azure resources, requiring a client layer to keep track, serialize and resume the operations through the lifecycle of the programs that wait for them to finish.

The [KeyVaultBackupClient][src-backup-client] offers three methods that execute long running operations:

- `beginBackup`, starts generating a backup of an Azure Key Vault on the specified Storage Blob account.
- `beginRestore`, starts restoring all key materials using the SAS token pointing to a previously stored Azure Blob storage backup folder.
- `beginSelectiveRestore`, starts restoring all key versions of a given key using user supplied SAS token pointing to a previously stored Azure Blob storage backup folder.

The methods that begin long running operations return a poller that allows you to wait indefinitely until the operation is complete. More information is available on the examples below.

## Examples

> We have provided readmes with the detailed steps to run our samples, both for [JavaScript](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/javascript/README.md) and [TypeScript](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/README.md). Direct links to the specific samples follow:

Reviewer: The link texts `JavaScript` and `TypeScript` do not have enough context when read through screenreaders

```suggestion
We have samples both in JavaScript and TypeScript that show the access control and backup/restore features in this package. Please follow the corresponding readmes for detailed steps to run the samples.
- [Readme for JavaScript samples](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/javascript/README.md)
- [Readme for TypeScript samples](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/README.md)

Direct links to the specific samples follow:
```

Author: Awesome feedback! thank you!

    - [Creating a Role Assignment](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/src/accessControlHelloWorld.ts)
    - [Getting a Role Assignment](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/src/accessControlHelloWorld.ts)
    - [Deleting a Role Assignment](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/src/accessControlHelloWorld.ts)
- Backup and restore:
    - [Performing a full key backup](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/src/backupRestoreHelloWorld.ts)
    - [Performing a full key restore](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/src/backupRestoreHelloWorld.ts)
    - [Performing a selective key backup](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/src/backupSelectiveRestore.ts)

>     - [Performing a selective key restore](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/samples/typescript/src/backupSelectiveRestore.ts)

Reviewer: All the above samples are pointing to the TS ones. Should we use JS ones instead? All TS users can be expected to know how to run JS code but not vice versa

Author: Sounds good! Can do

You can find more code samples through the following links:

- [KeyVault Admin Samples (JavaScript)](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/keyvault/keyvault-admin/samples/javascript)

> - [KeyVault Admin Samples (TypeScript)](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/keyvault/keyvault-admin/samples/typescript)src/

Reviewer 3: ```suggestion
- [KeyVault Admin Samples (TypeScript)](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/keyvault/keyvault-admin/samples/typescript)
```

- [KeyVault Admin Test Cases](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/keyvault/keyvault-admin/test/)

## Contributing

If you'd like to contribute to this library, please read the [contributing guide](https://github.com/Azure/azure-sdk-for-js/blob/master/CONTRIBUTING.md) to learn more about how to build and test the code.

![Impressions](https://azure-sdk-impressions.azurewebsites.net/api/impressions/azure-sdk-for-js%2Fsdk%2Fkeyvault%2Fkeyvault-admin%2FREADME.png)

<!-- LINKS -->
[NPM]: https://www.npmjs.com/
[package-gh]: https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/keyvault/keyvault-admin
[package-npm]: https://www.npmjs.com/package/@azure/keyvault-admin
[identity-npm]: https://www.npmjs.com/package/@azure/identity
[docs]: https://docs.microsoft.com/javascript/api/@azure/keyvault-admin

> [docs-service]: https://azure.microsoft.com/en-us/services/key-vault/

Reviewer 2: nit: remove "en-us" from url, don't want to pre-specify region

[azure-sub]: https://azure.microsoft.com/free/
[azure-cli]: https://docs.microsoft.com/cli/azure
[azure-identity]: https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/identity/identity
[api-rest]: https://docs.microsoft.com/rest/api/keyvault/
[compiler-options]: https://www.typescriptlang.org/docs/handbook/compiler-options.html
[dotenv]: https://www.npmjs.com/package/dotenv]
[src-rbac-client]: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/src/KeyVaultAccessControlClient.cs
[src-backup-client]: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/src/KeyVaultAccessControlClient.cs
[DAC]: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/identity/Azure.Identity/README.md
[storage-account-create-ps]: https://docs.microsoft.com/azure/storage/common/storage-quickstart-create-account?tabs=azure-powershell
[storage-account-create-cli]: https://docs.microsoft.com/azure/storage/common/storage-quickstart-create-account?tabs=azure-cli
[storage-account-create-portal]: https://docs.microsoft.com/azure/storage/common/storage-quickstart-create-account?tabs=azure-portal

> [package-gh]: https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/core/core-lro

Reviewer: ```suggestion
[core-lro]: https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/core/core-lro
```
